### DISCUSSION TITLE
C++ API officially not supported

### DISCUSSION THREAD

**Discussion by lkordos** [duckdb/duckdb#13407] [General]
Because the duckdb is written in C++, I would expect that C++ API would be a 1st class citizen. However, the C++ API for duckdb is not officially supported.

Could some of the duckdb devs bring a bit of light into this and explain why such a decision was made and what was the reasoning for it?

Also, is there any plan for the future, to support C++ API?

Thank you.

**Comment by Tishj** [comment]
The C++ API is absolutely a first class citizen, I don't know where it is said it's not officially supported?

Perhaps you mean that it's not stable, which is true. There is no stable C++ API, and it can be broken with every release, but it is by far the most feature rich API.

If you're looking for a stable API you should look towards the C API, as it's goal is to provide a stable wrapper on top of the volatile C++ API

**Comment by lkordos** [comment]
look at :

https://duckdblabs.com/news/2023/10/02/support-policy.html

Quote:

> The C++ API is intended for internal use and is not designed as a stable user-facing API.

That means other APIs have priority. That leads to my question, which I started this discussion with.

**Reply by Tishj** [reply]
"stable user-facing API"
I covered that, the answer is no, a stable C++ API is not planned currently

**Comment by Mytherin** [comment]
The current C++ API that exists is our internal API. While you are free to use this for extension development and similar, it is subject to change and has no guarantees of being stable. It exposes a much larger surface area than other APIs, and allows you to customize and access essentially everything in the system. Because this is our internal API, changes we make will make modifications to this API. Supporting this and stabilizing this is not feasible without crippling our ability to make changes to the system - as such we don't officially support this API. The surface area is simply too big. 

We have been thinking about building a separate, stable C++ API on top of the C API in the future. This will be separate from the current C++ API, and will be limited to what is possible to accomplish using the C API. Such an API would then be stable and officially supported.

**Reply by lkordos** [reply]
Thank you for explanation. I believe that C++ public, supported API would be greatly appreciated among C++ developers.
